sdk/src/main/scala/com/horizen/account/state/ForgerStakeMsgProcessor.scala

```diff
@@ -0,0 +1,709 @@
+  }
+
+  def getListItem(view: BaseAccountStateView, tip: Array[Byte]): (AccountForgingStakeInfo, Array[Byte]) = {
+    if (!linkedListNodeRefIsNull(tip)) {
+      val node = findLinkedListNode(view, tip).get
+      val stakeData = findStakeData(view, node.dataKey).get
+      val listItem = AccountForgingStakeInfo(
+        node.dataKey,
+        ForgerStakeData(
+          ForgerPublicKeys(
+            stakeData.forgerPublicKeys.blockSignPublicKey, stakeData.forgerPublicKeys.vrfPublicKey),
+          stakeData.ownerPublicKey, stakeData.stakedAmount)
+      )
+      val prevNodeKey = node.previousNodeKey
+      (listItem, prevNodeKey)
+    } else {
+      throw new ExecutionRevertedException("Tip has the null value, no list here")
+    }
+  }
+
+  def linkedListNodeRefIsNull(ref: Array[Byte]): Boolean =
+    BytesUtils.toHexString(ref).equals(BytesUtils.toHexString(LinkedListNullValue))
+
+  override def addScCreationForgerStake(msg: Message, view: BaseAccountStateView): Array[Byte] =
+    doAddNewStakeCmd(msg, view, isGenesisScCreation = true)
+
+  def doAddNewStakeCmd(msg: Message, view: BaseAccountStateView, isGenesisScCreation: Boolean = false): Array[Byte] = {
+
+    // check that message contains a nonce, in the context of RPC calls the nonce might be missing
+    if (msg.getNonce == null) {
+      throw new ExecutionRevertedException("Call must include a nonce")
+    }
+
+    // check that msg.value is greater than zero
+    if (msg.getValue.signum() <= 0) {
+      throw new ExecutionRevertedException("Value must not be zero")
+    }
+
+    // check that msg.value is a legal wei amount convertible to satoshis without any remainder
+    if (!isValidZenAmount(msg.getValue)) {
+      throw new ExecutionRevertedException(s"Value is not a legal wei amount: ${msg.getValue.toString()}")
+    }
+
+    val sender = msg.getFrom.address()
+    // check that sender account exists (unless we are staking in the sc creation phase)
+    if (!isGenesisScCreation && !view.accountExists(sender)) {
+      throw new ExecutionRevertedException(s"Sender account does not exist: ${msg.getFrom.toString}")
+    }
+
+    val inputParams = getArgumentsFromData(msg.getData)
+
+    val cmdInput = AddNewStakeCmdInputDecoder.decode(inputParams)
+    val blockSignPublicKey: PublicKey25519Proposition = cmdInput.forgerPublicKeys.blockSignPublicKey
+    val vrfPublicKey: VrfPublicKey = cmdInput.forgerPublicKeys.vrfPublicKey
+    val ownerAddress: AddressProposition = cmdInput.ownerAddress
+
+    if (!view.isEoaAccount(ownerAddress.address())) {
+      throw new ExecutionRevertedException(s"Owner account is not an EOA")
+    }
+
+    // TODO decide whether we need to check also genesis case (also UTXO model)
+    if (!isGenesisScCreation && networkParams.restrictForgers) {
+      // check that the delegation arguments satisfy the restricted list of forgers.
+      if (!networkParams.allowedForgersList.contains((blockSignPublicKey, vrfPublicKey))) {
```

Comment:
As a separate task: we should implement "open stake" feature like in open stake transaction for UTXO model.